### ISSUE TITLE
move python bindings packaging to main valhalla

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4624]
as it seems like I'll be taking a OSS sabbatical for a while and I'm trying to distribute my projects:)

we have https://github.com/gis-ops/pyvalhalla which only packages the bindings and pushes them to [PyPI](https://pypi.org/project/pyvalhalla/). the whole workflow is really messed up currently and I'd like to clean it up, so everything works with vcpkg & builds cleanly on CI.

to be honest, the whole linux business to publish c++-extended pypi packages is :vomiting_face: (osx & win is no problem) because of ABI compat with python, packages for distribution have to be compiled on distros with a glibc version as low as possible for maximum distro compatibility. I set up https://github.com/gis-ops/manylinux, which is a forked manylinux image (based on almalinux) with glibc 2.28 also installing all necessary dependencies, which should be good for another few years before being deprecated.

I'd propose to add the workflows etc to our main repo which get triggered when there was a tag/release created. would you be fine with that? there's very little extra files needed, mainly the `setup.py` and the GHA workflow yaml.

**Comment by nilsnolde** [comment]
and possibly we could fork the manylinux image to the org, similar as I did. it's not much changed from upstream, mostly installing valhalla dependencies and preserving python static libraries as we need them for compilation.

**Comment by thomthom** [comment]
I'm also curious about this. I've been using the pyvalhalla bindings, which currently as of the latest version `3.2.0` supports up to python 3.11, using manylinux wheels and no source distro.

I can understand the no source distro thing, after having fumbling around to get a python C extension package published that made use of C++23. (I'm mainly using Windows, so it was a learning curve to how dependencies work on Linux.)

I'm curious, could the existing workflow be setup in this repo? And wire up publishing based on new releases, like Nils suggests? Or are there things that needs to be refactored first? 

> so everything works with vcpkg & builds cleanly on CI.

Is the current state such that the python bindings aren't building on CI?

**Comment by nilsnolde** [comment]
Yeah, the really messed up thing was that all the dependency binaries and headers were checked into the repo for all platforms. Win was hard to get right, since the vcpkg dependencies created such huge artifacts, it exceeded the free limit for caching.. and it takes 45 mins to build the deps from source. And that for every commit. Only to build the Valhalla library, which almost never changes bcs we went with releases. It seemed a huge waste, big enough to build the deps on each platform manually and check them into the repo. But it's likely that vcpkg improved a lot since then and it might now be able to build and cache only release binaries so the size becomes manageable. Also I think at the start there were no OSX runners on GH. 

That's all to say that I think it should be doable and much less resource wasting by now, but it'll definitely be a lot of CI hell with Windows & vcpkg. I'd like to look into it again, but won't promise anything;)

**Comment by nilsnolde** [comment]
Great GH handle by the way😄 TomTom's people usually have the -tomtom suffix, so you'd be ThomThom-tomtom if you'd work there haha

**Comment by nilsnolde** [comment]
Just leaving a few thoughts here.

The whole process is just about packaging the bindings, we're already building them in this repo on CI for all platforms. For OSX & Win even on GHA. That means we can use the already compiled pyvalhalla package for packaging for those 2 platforms and we don't really have to care about rebuilding dependencies, or even valhalla for that matter. If we do that, then one disadvantage is that we still do need the same dependency .so's for packaging as the Valhalla libraries used to be compiled with. The wheel not only contains the compiled valhalla library but also all the .so files of all dependencies. Which kinda defeats the utility of shared libraries.. Not sure how to compile a fully _static_ python extension module. Actually pyvalhalla doesn't need any of `mjolnir` (yet, though I thought about exposing the build process too). For Win GHA, we could probably load from the same `vcpkg` cache as the existing workflow and other platforms have real package managers.

For Linux the packaging process is a LOT more involved and our current circle pyvalhalla binaries are sadly worthless for PyPI (see below), but then we also don't have to think about how to transfer stuff from circleci - > GHA. The `manylinux` docker image and the `cibuildwheel` GHA make things at least tolerable. The `manylinux` image hosts a pretty old distro with an old `glibc` version, old enough to be supported by most modern distros. So that image is used to build the bindings, including rebuilding Valhalla (if sources changed). When producing that image, it's best to already install Valhalla's dependencies so they're not being rebuilt every time (some might have to be built from source, e.g. protobuf). We should also be able to cache the `.ccache` folder for Valhalla to speed things up.

 I'd suggest we:

- fork the `manylinux` repo to our Valhalla org and maintain our own Linux build image (mostly installing Valhalla's dependencies), it's super easy
- only release the bindings on valhalla releases to the existing pyvalhalla PyPI package
- build a Linux wheel for a PR, since Linux uses another old Linux distro (incl old dependency versions) for packaging
- perhaps think about having a `pyvalhalla-git` package which tracks `master` due to our fairly irregular release process
- we could also widen our supported architecture spectrum: I saw GHA now has `linux/win-arm64` in ["public preview"](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories). I'd suggest we only add Linux ARM (and probably switch the current one to GHA). WinARM can be added if someone really wants it IMO.

**Comment by nilsnolde** [comment]
Hm, I just remembered smth else. Originally, pyvalhalla was built with a `setup.py`, it didn't rely on the normal build of the bindings which is done in CMake. I can't remember what the problem was with using the CMake ones. I think it was just so that I could use e.g. `python3.10 setup.py bdist_wheel` to create the wheel. And maybe turning on `PIC` for the static valhalla library (possibly we could use the dynamic one?). To build a wheel we could do [this](https://medium.com/@jonathan_b/building-a-python-wheel-with-pyproject-toml-bbab842e989a) with a `pyproject.toml`, which is a lot nicer to configure than a `setup.py`. Maybe `PIC` can be turned on for the existing CI artifacts from OSX & Win (?), Linux has to be built separately anyways.

Makes me hopeful to finally ditch the awful current setup:)

**Comment by thomthom** [comment]
Thank you for the detailed feedback. I need to come back to these when I'm home from my travels and have my dev machine at hand to test things.

> But it's likely that vcpkg improved a lot since then and it might now be able to build and cache only release binaries so the size becomes manageable.

hm.. good question. I have used vcpkg only for one personal project. Though it didn't have many dependencies so I didn't look into how the workflow cached the build. What kind of size limit did you run into? GitHub workflows?

> Great GH handle by the way😄 TomTom's people usually have the -tomtom suffix, so you'd be ThomThom-tomtom if you'd work there haha

There's been a few people who have assumed I work at TomTom. I will at time sit in the back seat and provide turn by turn directions. 😄

> If we do that, then one disadvantage is that we still do need the same dependency .so's for packaging as the Valhalla libraries used to be compiled with.

Isn't pyvalhalla built against static valhalla?

I'm pretty green to python C extensions, only made one, but then linking dependencies statically to avoid having to deal with how/where dynamic libraries are located/loaded. Made a few Ruby C extensions, did the same thing there. On Windows I even linked against static CRT to make it completely standalone - sans its dependency against Ruby.

I have seen that Python offer some kind of Stable C API: https://docs.python.org/3/c-api/stable.html
Not sure if that is something pyvalhalla can use nor if it makes anything any simpler... I've just read about it but not dived into it.

> Hm, I just remembered smth else. Originally, pyvalhalla was built with a `setup.py`, it didn't rely on the normal build of the bindings which is done in CMake. I can't remember what the problem was with using the CMake ones. I

Yea, this was puzzling me. I see a warning in the root CMakeLists.txt:

https://github.com/valhalla/valhalla/blob/346e9c619efae18a2a99ee4db115f2f09a06aeac/CMakeLists.txt#L5-L8

However, it's not clear to me what/how the offical valhalla build is done if not with CMake. The GH CI seems to use CMake. So does the Docker image. And the docs says: https://valhalla.github.io/valhalla/building/

> Valhalla uses [CMake](https://cmake.org/) as the build system, the compiler needs C++17 support.

So I'm wondering if the warning the CMakeLists.txt file is out of date?

And while on C++ support, the docs says C++17, is this an enforced limit in the code base? Wondering since manylinux seems to be lagging way behind in what C++ version it allows. Or is that a moot issue if using a custom manylinux build?

**Comment by thomthom** [comment]
Btw, the valhalla docker image, does it come with python and the pyvalhalla bindings? Could that potentially be used as a base for a python app that uses pyvalhalla and possibly making it easier to stay up to date with releases?

**Comment by nilsnolde** [comment]
> Isn't pyvalhalla built against static valhalla?

It is and admittedly my in-depth knowledge of the compilation & linking process is quite thin here, but the final .so for the bindings is just 11 MB and if you run `ldd` on it, it depends on the .so's from valhalla dependencies.

> So I'm wondering if the warning the CMakeLists.txt file is out of date?

Oh wow, yeah, that's massively out of date. It's all CMake now and the old way was abandoned a longer time ago. 

> And while on C++ support, the docs says C++17, is this an enforced limit in the code base? Wondering since manylinux seems to be lagging way behind in what C++ version it allows. Or is that a moot issue if using a custom manylinux build?

`manylinux` bases off of AlmaLinux 8 and states that it has `gcc v14.x` for [the image](https://github.com/pypa/manylinux?tab=readme-ov-file#manylinux_2_28-almalinux-8-based) we'd be using.  Even the native gcc 8.5 has full C++17 support, but the custom-built gcc version is even fit for C++20.

> Btw, the valhalla docker image, does it come with python and the pyvalhalla bindings? Could that potentially be used as a base for a python app that uses pyvalhalla and possibly making it easier to stay up to date with releases?

That's a very good point. Yes, it could, the python bindings are installed to the system python and work without issues: https://github.com/valhalla/valhalla/blob/56ed70e183ddd86304d6e4ce86a057a325a86b3b/Dockerfile#L80

**Comment by nilsnolde** [comment]
FWIW, this is WIP now, I'll keep a log of things that work already here:
- created a new repo https://github.com/valhalla/manylinux, which contains some fixes & additions to the the upstream one, mostly building Valhalla dependencies from source
- linux bindings build now within the `manylinux` image above

TODO:
- ~~try linux build with `cibuildwheel` and publish to testing pypi, then install pyvalhalla in a fresh docker container~~ ✔ 
- ~~take care of other platforms~~ ✔ 
- ~~try to transfer the pypi package to some shared valhalla user~~ don't need to, auth is done automagically
- ~~create a `pyvalhalla-git` pypi package to track master~~ ✔

**Comment by nilsnolde** [comment]
Yay, the Linux x64 bindings work now for all minor python versions: https://test.pypi.org/project/pyvalhalla/3.5.1.post137/. They're published from GitHub actions, takes ~ 15 mins for the whole workflow.

**Comment by nilsnolde** [comment]
Argh.. I thought OSX would be easy, but I forgot that we can't work with `brew` versions as those are compiled to work with the host's OSX version building the bindings, while the python wheels should work with OSX versions down to 11.0 or so. It's unacceptable to (recursively) build all `mjolnir` dependencies from source. So either, we:
1. accept that only OSX v14 (arm)/OSX v13(intel) or higher can use the PyPI bindings or
2. we install another `libvalhalla` with `DENABLE_DATA_TOOLS=OFF` (and `-DENABLE_SERVICES=OFF` etc), just to build the bindings on CI, in which case we'd only need to build protobuf from source
3. use vcpkg to install & cache OSX dependencies (need to figure out how to tell vcpkg to compile for older OSX versions..)

1 would be the laziest approach, but would likely work for > 95% of OSX users. 2. would be easy to do, but the problem would re-surface once we decide to expose build tools to the bindings. 3. is then the preferred long-term solution.

Since I have no intention to whack around vcpkg on OSX, I'd actually go with 2. BUT, IMO 1. is preferable here. That way chances are higher that someone really really wants the bindings to work on their old OSX version and we can tell them then to put in the work for 3 hehe

EDIT: quick search and 3. seems to be infeasible for now: https://github.com/microsoft/vcpkg/issues/10038

**Comment by thomthom** [comment]
> Yay, the Linux x64 bindings work now for all minor python versions: https://test.pypi.org/project/pyvalhalla/3.5.1.post137/. They're published from GitHub actions, takes ~ 15 mins for the whole workflow.

Awesome! That will simplify my docker dependency constraints. I've had to avoid the newest ubuntu images as they had py 3.12 installed.



> 1. accept that only OSX v14 (arm)/OSX v13(intel) or higher can use the PyPI bindings or

I'm not super familiar with macOS, but I get the impression that the upgrade rate on macOS is pretty high. I usually hold back on major macOS versions to let (in)compatibilities settle before I upgrade, but I'm on Sonoma on both my dev macs.

https://telemetrydeck.com/survey/apple/macOS/versions/ <- seems to suggest a pretty evergreen environment.

Option 1 seems like at least a good start. Additional support can be added if there's asks for it?

**Comment by nilsnolde** [comment]
Right, it can be done incrementally. I guess most osx users behave the same as you, running 1 version behind. I'd just go with whatever GHA has as oldest.

**Comment by nilsnolde** [comment]
@thomthom if you have a chance you can try the OSX wheels, see last sentence of PR description in https://github.com/valhalla/valhalla/pull/5280.

**Comment by thomthom** [comment]
> [@thomthom](https://github.com/thomthom) if you have a chance you can try the OSX wheels, see last sentence of PR description in [#5280](https://github.com/valhalla/valhalla/pull/5280).

https://github.com/valhalla/valhalla/pull/5280#issuecomment-2944935257

**Comment by nilsnolde** [comment]
@thomthom I remember you asked if it's worth it to contribute typing for pyvalhalla, but I can't seem to find the comment anymore (did you delete the comment?). I'd love to have that! I didn't fully catch on to typing in Python tbh, I didn't really do serious Python stuff in a while, e.g. I never used `mypy`. But do note, we're generally fine including well maintained external dependencies like `pydantic`, for `mypy` I'd first need to understand what it really is that is does.

I think it's not necessary to provide an extra repo only for the stubs, but again, not sure about compatibility across minor versions. We generally package for all [officially supported](https://devguide.python.org/versions/) CPython minor versions, e.g. in a few months we update to 3.14 and will drop 3.9. Or more specifically, we package all versions which are supported for the [`cibuildwheel`](https://github.com/pypa/cibuildwheel) version we're using, they might have some grace period for unsupported versions.

**Comment by thomthom** [comment]
I made that comment in the PR you have open for the python bindings: https://github.com/valhalla/valhalla/pull/5280#issuecomment-2944935257

I'll have a closer look at typing support in Python 3.9/3.10. I've mostly used 3.11+ and 3.11 had a few improvements to the typing system.

Mainly I add type hints inline. But if that complicates things for older python versions then maybe a separate stub can work until the oldest straggling versions are dropped.

I also have planned a PR for JSON data clarifications in the API docs. I keep getting confused, and even AI tools get confused by the docs on that is input vs output params. This would tie into how I'm mapping out the valhalla data type into typed dictionaries - my prototype already caught a bunch of potential mistakes/bugs I'd made in accessing the data (or crafting the input data).

I have also recently stated poking into pydantic which seems to be even nicer to work with then typed dictionaries. Not sure if valhalla want to start consuming that instead. Or maybe rely on the convertion to/from dict to pydantic types. Pydantic types is much more IDE friendly in terms of auto-complete. This would all be an API change, so probably less likely to consume it instead of a dict. But... overload? (This can also be another discussion later after getting typed dictionaries defined).

**Comment by nilsnolde** [comment]
> Mainly I add type hints inline. But if that complicates things for older python versions then maybe a separate stub can work until the oldest straggling versions are dropped.

Hm, maybe we could live with the shortcomings of 3.9/3.10 for the time being and enable whatever improvements once we drop those in 4 - 16 months? Again, I have very little idea what the impact would be for typing support. That's assuming the separate stubs live in another repository and is distributed to a separate package, which would involve quite a bit of additional setup & maintenance. But maybe it's also possible to cleverly import and declare types based on the Python minor version? Or maybe packages like `pydantic` or other external libs provide API for just that? I'm sure they figured already out how to make types minor-version-agnostic.

> I also have planned a PR for JSON data clarifications in the API docs. I keep getting confused, and even AI tools get confused by the docs on that is input vs output params. This would tie into how I'm mapping out the valhalla data type into typed dictionaries - my prototype already caught a bunch of potential mistakes/bugs I'd made in accessing the data (or crafting the input data).

What does `I also have planned a PR for JSON data clarifications in the API docs` imply? You mean simply re-organizing the markdown docs so autodiscovery for IDEs work better? That'd be very nice if it's still easily human-readable. Wasn't there also some semi-standard way to provide IDEs with autocomplete for JSON files like our config file? I remember smth like a `"$scheme"` key or so pointing to some remote JSON doc which details the options for each key and their meaning, so e.g. VS Code would show them to the user..

Basically IMO whatever improves usability is very worthy doing and we'd be grateful for contributors to take those on:) My own motivation is not that super high, bcs I'll hardly use those bindings myself. Always happy to review those PRs and learn smth along the way though!

You're right, let's continue that discussion once you found the time & energy to raise a PR. Or feel free to open an issue about it to continue this discussion first, e.g. whether a separate stubs project or we can keep it internal. And thanks for your involvement here, I'm very happy there's someone else having an interest in improving these things (that is not to be read as passive-aggressiveness towards other maintainers or contributors 🥰 ).

**Comment by thomthom** [comment]
> What does `I also have planned a PR for JSON data clarifications in the API docs` imply? You mean simply re-organizing the markdown docs so autodiscovery for IDEs work better?

I was thinking with human readability in mind. The IDE comment was just an observation that GPT like agents struggled to provide proper answers from the docs. But humans first, always.

> And thanks for your involvement here, I'm very happy there's someone else having an interest in improving these things (that is not to be read as passive-aggressiveness towards other maintainers or contributors 🥰 ).

Happy to help. :)
If this project I'm tinkering with ends up with a working POC then it's good to have some involvement and insight to the code base. Starting with doc fixes and typehint fixes is an easy entry to getting familiar.

**Comment by nilsnolde** [comment]
Sounds good! We're happy with anyone who relies on our stuff, much more so if that earns us contributors:) Let us know of any other hurdles or questions in the routing space, even if it fails it'd great to know why. Maybe smth can be improved. The thing is that some of us (maintainers) hardly use our own software beyond developing (except @dnesbitt61 & @gknisely), which is really not ideal. We rely much on those who actually do use it to report issues and ideally offer fixes.

Also note, that @chrstnbwnkl offers commercial support in case that should become interesting or your problem exceeds the effort of what the community can be expected to provide (e.g. implementing external traffic sources).

**Comment by thomthom** [comment]
> Sounds good! We're happy with anyone who relies on our stuff, much more so if that earns us contributors:) Let us know of any other hurdles or questions in the routing space, even if it fails it'd great to know why. Maybe smth can be improved. The thing is that some of us (maintainers) hardly use our own software beyond developing (except [@dnesbitt61](https://github.com/dnesbitt61) & [@gknisely](https://github.com/gknisely)), which is really not ideal. We rely much on those who actually do use it to report issues and ideally offer fixes.

Appreciate that.

So far valhalla has been working well out of the box.

Though I've had some challenges with getting the tile data. Like this one for instance; https://github.com/valhalla/valhalla/issues/4729

The valhalla server routed fine, but not my own python app. I've not dug into what's causing it yet. That's one of the reasons I'm trying to get it building on Windows, so I can step through in the debugger and snoop around.

The other main unknown is deploying my own valhalla server. The infrastructure, and figuring out how to scale it. How to set up regions (which seems to be key...). I'm hoping to deploy a test server soon to learn more about that, or at least learn more about how much I don't know. I really don't want to be dealing with hosting to be honest, and Mapbox would probably get me a long way. But there are some key functionality I want to get out of this that I can only get via the route attributes, inspecting the road data. Doesn't seem that anyone provide anything similar. (Despite Mapbox using Valhalla, I think? Or they acquired the mapzen team?)

> Also note, that [@chrstnbwnkl](https://github.com/chrstnbwnkl) offers commercial support in case that should become interesting or your problem exceeds the effort of what the community can be expected to provide (e.g. implementing external traffic sources).

Yes! That would be very interesting. I've tried to look for commercial support/consulting services. But it all appeared to dead-end at old mapzen links.

**Comment by thomthom** [comment]
Maybe I should introduce myself a bit, given that I've been nosing around for a while now.

I stumbled across valhalla because I was frustrated with the mapping tools available when I was planning and tracking my motorcycle trips. Nothing quite fit my needs/wishes. I felt I had to do a lot of work, mainly data handling, that really should be a computer's job. So, I defaulted to what I so often do; start writing my own tool. Got a crude flask app up and running early last year, which got me through more than 8000km around Scandinavia, mostly Norway.

- https://www.thomthom.net/thoughts/2024/09/motorcycle-trip-from-trondheim-norway-to-lund-sweden-and-back-via-denmark-may-2024/
- https://www.thomthom.net/thoughts/2024/12/2024-summer-motorcycle-trip-around-dovre/
- https://www.thomthom.net/thoughts/2024/12/the-great-motorcycle-road-trip-of-2024/

Then over the winter (short motorcycle seasons this far up north, and long dark winters) I continued working on the tool. It's been mostly storing things locally on my machine so far. But I've been working on setting up the infrastructure for a login system, PostGIS to persist data. And next step is getting a POC deployed somewhere. I'd hoped I'd get to that already this year, and spend the summer testing what I got, collecting data. Then spend the next winter to get something that other people can also use.

**Comment by thomthom** [comment]
hm.. maybe this should have been split off into the Discussion section... 🤷‍♂️

**Comment by nilsnolde** [comment]
> The other main unknown is deploying my own valhalla server. The infrastructure, and figuring out how to scale it. How to set up regions (which seems to be key...). I'm hoping to deploy a test server soon to learn more about that, or at least learn more about how much I don't know. I really don't want to be dealing with hosting to be honest, and Mapbox would probably get me a long way. But there are some key functionality I want to get out of this that I can only get via the route attributes, inspecting the road data. Doesn't seem that anyone provide anything similar. (Despite Mapbox using Valhalla, I think? Or they acquired the mapzen team?)

It's not hard to host once it's set up and is running on an appropriate machine. Our public instance is running since 3 years or so and I maybe had to intervene 5-6 times in that period, which was always because of some new library or smth other super obvious. We'll likely re-do our setup this year as we're probably moving servers. Then it should be `ansible` powered and easy to reproduce. At that point I think it makes sense to open-source it. Currently it's not the best setup, I wouldn't like to have that in the open and get issues for that crap;)

But Mapbox is of course also an option 👼 

I know GraphHopper also has a pretty enthusiastic motorbiker (ex-?)maintainer, running https://kurviger.com. But I'm not sure kurviger's additions to GraphHopper are all open (kinda sure they're not).

**Comment by thomthom** [comment]
Is the public valhalla instance running on a single machine? With the entire world tileset? Any stats on what machineware and number of requests/performance it runs? And where? In a cloud provider or self hosting?

The hardware needed to valhalla work in a Release product environment is the biggest questionmark for me right now.

I see that using attributes to obtain route details is slowing things down noticeable, so I might need to look into doing fetching that info async from the actual routing.

**Comment by nilsnolde** [comment]
The service is running the whole world on a single machine with 64 GB RAM & 16 threads at https://hetzner.com. Sadly we don't have service-level stats running, only monitoring health. 

Mainly you:
- _ideally_ want enough RAM to read the full tileset, but it's fine to have it smaller, especially if there's focus areas; note, RAM is also needed for the algorithms, leave plenty for that (depending on length of route(s) and which service (matrix is expensive))
- want the fastest SSD you can get, we're reading a lot from disk at runtime
- enough threads to handle the service; it's best if you do your own bit of load testing

You can run `valhalla_service` in the beginning and when in prod you can switch to the more robust setup of multiprocessing the individual workers for `loki`, `thor` & `odin` and using some process manager, once you got more comfortable with the project.

Depending on how serious you want to get, of course hosting such a service can be a huge pain. Just don't do OEM's and you could be fine haha

**Comment by nilsnolde** [comment]
ah this can actually be closed!

**Comment by thomthom** [comment]
> The service is running the whole world on a single machine with 64 GB RAM & 16 threads at https://hetzner.com. Sadly we don't have service-level stats running, only monitoring health.

hm.. interesting. Any rough notion of the amount of traffic that server gets? 

> Mainly you:
> 
>     * _ideally_ want enough RAM to read the full tileset, but it's fine to have it smaller, especially if there's focus areas; note, RAM is also needed for the algorithms, leave plenty for that (depending on length of route(s) and which service (matrix is expensive))

Ah, I thought CPU speed was key factor. 

>     * want the fastest SSD you can get, we're reading a lot from disk at runtime

That's good to know. I was thinking of trying out attached storages, so that I could keep the tileset data separate from the node.

>     * enough threads to handle the service; it's best if you do your own bit of load testing
> 
> You can run `valhalla_service` in the beginning and when in prod you can switch to the more robust setup of multiprocessing the individual workers for `loki`, `thor` & `odin` and using some process manager, once you got more comfortable with the project.

I though I would have to have multiple nodes for different regions to keep things fast enough. Locally I've been testing on a tileset covering only Scandinavia. I was going to fire up a server with that data set and make some measurements. Maybe I don't need that complexity after all. In any case I was planning on keeping the routing server separate from the web app/site - communicating via local network between the nodes. If anything would need scaling I'd expect that to be the routing backend.

> Depending on how serious you want to get, of course hosting such a service can be a huge pain. Just don't do OEM's and you could be fine haha

"OEM"?

**Comment by nilsnolde** [comment]
> hm.. interesting. Any rough notion of the amount of traffic that server gets?

As I said, sadly no service-level stats. I totally forgot about it, but we did have an ELK stack running actually.. I just checked it out, only getting 

![Image](https://github.com/user-attachments/assets/d3fe7950-1d27-499c-8b05-1eba187987ee)

I remember years ago, when I last touched that bit (didn't set it up myself initially), it used to break down regularly for some reason I can't remember. But our monitoring would bark if we reached >=80% of the RAM capacity IIRC.

> Ah, I thought CPU speed was key factor.

CPU definitely is super important, maybe more so than RAM, but more important are fast disks.

> I though I would have to have multiple nodes for different regions to keep things fast enough.

If you're talking AWS regions, then yeah, good idea:) However, if you're talking geographic regions, there's no need. We're loading graph tiles dynamically as they occur in the trip expansion, and by default RAM is managed by the OS bcs we're mmapping. Most other OSS routing engines work with a RAM monolith, e.g. OSRM, where breaking up regions makes (or used to make) sense to keep RAM requirements down for a single machine (they need > 500 GB for the planet I think). 

> In any case I was planning on keeping the routing server separate from the web app/site - communicating via local network between the nodes. If anything would need scaling I'd expect that to be the routing backend.

Yep, I'd keep them separate too and agree with the scaling thought. Do note, you'll need a separate machine to build the tiles. For the public we're simply building endlessly, as soon as one planet finished, the next one starts. For most needs a weekly build should be more than fine, though it's of course nice to have OSM updates as they roll in:)
